### DISCUSSION TITLE
The early stopping Plugin parameterization

### DISCUSSION THREAD

**Discussion by AymenTlili131** [ContinualAI/avalanche#1382] [Q&A]
Hi , I'm trying to train multiple models each taking a lengthy amount of time if they were to go through all the epoch . I'd like them to stop when convergence is achieved (I'll describe this as not improving by more than 0.05% in terms of accuracy) 
so to me accuracy at epoch e: 95.00% and epoch e+1:  95.04% are the same and I'd like to exit at epoch e+1 while also checkpointing .
wanted to ask if the following parametrization is correct when working on 2 experiments of the splitMnist Benchmark .
```
ES=EarlyStoppingPlugin(        
        patience=0,
        val_stream_name="test_stream",
        metric_name= "Top1_Acc_Epoch",
        mode= "max",
        peval_mode="epoch",
        margin=0.2,
        verbose=True,)
```
for a more wholistic view and code to test with  (note the if statement limiting this to one init/activation combination

```
import torch
import torchvision

from itertools import product
from itertools import combinations

from avalanche.benchmarks.classic import SplitMNIST

from avalanche.benchmarks.generators import dataset_benchmark


import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader
from torch.utils.tensorboard import SummaryWriter
import numpy as np
import torch.nn.functional as F
from torch.optim import Adam
from torch.nn import CrossEntropyLoss

from avalanche.training.supervised import Cumulative

from avalanche.training.plugins import EvaluationPlugin,EarlyStoppingPlugin
from avalanche.training.determinism.rng_manager import RNGManager
from avalanche.training.plugins.checkpoint import CheckpointPlugin, \
    FileSystemCheckpointStorage


from avalanche.evaluation.metrics import Accuracy, TaskAwareAccuracy
from avalanche.evaluation.metrics import accuracy_metrics
from avalanche.logging import InteractiveLogger,TensorboardLogger


import os
import json
import pandas as pd


class CNN(nn.Module):
    def __init__(
        self,
        channels_in,
        nlin="leakyrelu",
        dropout=0.0,
        init_type="uniform",
    ):
        super().__init__()
        # init module list
        self.module_list = nn.ModuleList()
        ### ASSUMES 28x28 image size
        ## compose layer 1
        self.module_list.append(nn.Conv2d(channels_in, 8, 5))
        self.module_list.append(nn.MaxPool2d(2, 2))
        self.module_list.append(self.get_nonlin(nlin))
        # apply dropout
        if dropout > 0:
            self.module_list.append(nn.Dropout(dropout))
        ## compose layer 2
        self.module_list.append(nn.Conv2d(8, 6, 5))
        self.module_list.append(nn.MaxPool2d(2, 2))
        self.module_list.append(self.get_nonlin(nlin))
        ## add dropout
        if dropout > 0:
            self.module_list.append(nn.Dropout(dropout))
        ## compose layer 3
        self.module_list.append(nn.Conv2d(6, 4, 2))
        self.module_list.append(self.get_nonlin(nlin))
        ## add flatten layer
        self.module_list.append(nn.Flatten())
        ## add linear layer 1
        self.module_list.append(nn.Linear(3 * 3 * 4, 20))
        self.module_list.append(self.get_nonlin(nlin))
        ## add dropout
        if dropout > 0:
            self.module_list.append(nn.Dropout(dropout))
        ## add linear layer 1
        self.module_list.append(nn.Linear(20, 10))

        ### initialize weights with se methods
        self.initialize_weights(init_type)

    def initialize_weights(self, init_type):
        # print("initialze model")
        for m in self.module_list:
            if type(m) == nn.Linear or type(m) == nn.Conv2d:
                if init_type == "xavier_uniform":
                    torch.nn.init.xavier_uniform_(m.weight)
                if init_type == "xavier_normal":
                    torch.nn.init.xavier_normal_(m.weight)
                if init_type == "uniform":
                    torch.nn.init.uniform_(m.weight)
                if init_type == "normal":
                    torch.nn.init.normal_(m.weight)
                if init_type == "kaiming_normal":
                    torch.nn.init.kaiming_normal_(m.weight)
                if init_type == "kaiming_uniform":
                    torch.nn.init.kaiming_uniform_(m.weight)
                # set bias to some small non-zero value
                m.bias.data.fill_(0.01)

    def get_nonlin(self, nlin):
        # apply nonlinearity
        if nlin == "leakyrelu":
            return nn.LeakyReLU()
        if nlin == "relu":
            return nn.ReLU()
        if nlin == "tanh":
            return nn.Tanh()
        if nlin == "sigmoid":
            return nn.Sigmoid()
        if nlin == "silu":
            return nn.SiLU()
        if nlin == "gelu":
            return nn.GELU()
        if nlin == "ORU":
            return ORU()
        if nlin == "ERU":
            return ERU()

    def forward(self, x):
        # forward prop through module_list
        for layer in self.module_list:
            x = layer(x)
        return x

    def forward_activations(self, x):
        # forward prop through module_list
        activations = []
        for layer in self.module_list:
            x = layer(x)
            if (
                isinstance(layer, nn.Tanh)
                or isinstance(layer, nn.Sigmoid)
                or isinstance(layer, nn.ReLU)
                or isinstance(layer, nn.LeakyReLU)
                or isinstance(layer, nn.SiLU)
                or isinstance(layer, nn.GELU)
                or isinstance(layer, ORU)
                or isinstance(layer, ERU)
            ):
                activations.append(x)
        return x, activations
    

RNGManager.set_random_seeds(1234)

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

eval_plugin = EvaluationPlugin(
    accuracy_metrics(minibatch=False, epoch=True, experience=True, stream=True),
    #loss_metrics(minibatch=False, epoch=True, experience=True, stream=True),
    #timing_metrics(epoch=True),
    #forgetting_metrics(experience=True, stream=True),
    #cpu_usage_metrics(experience=True),
    #confusion_matrix_metrics(num_classes=split_mnist.n_classes, save_image=False, stream=True),
    #disk_usage_metrics(minibatch=False, epoch=True, experience=True, stream=True),
    loggers=[InteractiveLogger(),TensorboardLogger()],
    strict_checks=False
)
ES=EarlyStoppingPlugin(        
        patience=0,
        val_stream_name="test_stream",
        metric_name= "Top1_Acc_Epoch",
        mode= "max",
        peval_mode="epoch",
        margin=0.0005,
        verbose=True,)




results = pd.DataFrame(columns=["index",'Strategy','seed','activation','init','model #','Dataset 1','Acc D1',"Dataset 2","Acc D2"])
S=combinations(range(10), 5)
All=list(range(10))
ListExperiences=list()
for i in S :
    L1=list(i)
    L2=[k for k in All if k not in L1] 
    L1.extend(L2)
    ListExperiences.append(L1)
c=0

L_activations=["relu","silu"]
#,"relu","tanh","silu","gelu"]
L_inits=["kaiming_normal","xavier_normal"]
#"xavier_uniform","uniform","normal","kaiming_uniform"]

for class_order_idx in range(len(ListExperiences)):
    split_mnist = SplitMNIST(n_experiences=2, fixed_class_order=ListExperiences[class_order_idx],shuffle=False)

    for i in product(L_activations,L_inits): #split this into 2 for loops #reduce number of activations
        print(list(i))
        model = CNN(1,nlin=list(i)[0],init_type=list(i)[1])
        model=model.to(device)

        optimizer = Adam(model.parameters(), lr=0.05)
        scheduler = torch.optim.lr_scheduler.CyclicLR(optimizer, base_lr=1e-3, max_lr=0.1, step_size_up=1, mode="triangular2", cycle_momentum=False)


        checkpoint_pluginCum = CheckpointPlugin(
            FileSystemCheckpointStorage(
                directory='./checkpoints/Cumulative/{}/{}/{}'.format(ListExperiences[class_order_idx][:5],list(i)[0],list(i)[1]),
            ),
            map_location=device
        )


        criterion = CrossEntropyLoss()
        cl_strategy = Cumulative(
            model, optimizer, criterion,
            train_mb_size=150, train_epochs=5, eval_mb_size=150, evaluator=eval_plugin,eval_every=1,device=device,plugins=[ES,checkpoint_pluginCum]
        )


        print('Starting experiment...')
        res = []

        print(len(split_mnist.train_stream))
        for experience_idx in range(len(split_mnist.train_stream)):
            print("Start of experience: ", split_mnist.train_stream[experience_idx].current_experience)
            print("Current Classes: ", split_mnist.train_stream[experience_idx].classes_in_this_experience)

            cl_strategy.train(split_mnist.train_stream[experience_idx])
            print('Training completed')

            # assume you have the following dictionary
            print('Computing accuracy on the whole test set')
            res.append(cl_strategy.eval(split_mnist.test_stream))
        for experience_idx in range(len(split_mnist.train_stream)):
            results.at[c,"Acc D1"]=res[experience_idx]['Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000']
            results.at[c,"Acc D2"]=res[experience_idx]['Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp001']
            results.at[c,'index']=c
            results.at[c,'Strategy']='Cumulative'
            results.at[c,'seed']=1234
            results.at[c,'activation']=list(i)[0]
            results.at[c,'init']=list(i)[1]
            results.at[c,'model #']=experience_idx
            results.at[c,'Dataset 1']=ListExperiences[class_order_idx][:5]
            results.at[c,'Dataset 2']=ListExperiences[class_order_idx][5:]
            c+=1
            #path=Path('./checkpoints/Cumulative/{}/{}/{}/{}/metrics.json'.format(ListExperiences[class_order_idx][:5],list(i)[0],list(i)[1],experience_idx+1))
            json_str = json.dumps(res[0], indent=4) + '\n'
            #path.write_text(json_str, encoding='utf-8')
            with open('./checkpoints/Cumulative/{}/{}/{}/{}/metrics.json'.format(ListExperiences[class_order_idx][:5],list(i)[0],list(i)[1],experience_idx+1), 'w') as outfile:
                outfile.write(json_str)


            np.save('./checkpoints/Cumulative/{}/{}/{}/{}/Top1_Acc_Epoch'.format(ListExperiences[class_order_idx][:5],list(i)[0],list(i)[1],experience_idx+1),eval_plugin.get_all_metrics()['Top1_Acc_Epoch/train_phase/train_stream/Task000'][1])
            np.save('./checkpoints/Cumulative/{}/{}/{}/{}/Top1_Acc_Epoch_indices'.format(ListExperiences[class_order_idx][:5],list(i)[0],list(i)[1],experience_idx+1),eval_plugin.get_all_metrics()['Top1_Acc_Epoch/train_phase/train_stream/Task000'][0])
            np.save('./checkpoints/Cumulative/{}/{}/{}/{}/Top1_Acc_Exp_0'.format(ListExperiences[class_order_idx][:5],list(i)[0],list(i)[1],experience_idx+1),eval_plugin.get_all_metrics()['Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000'][1])
            np.save('./checkpoints/Cumulative/{}/{}/{}/{}/Top1_Acc_Exp_0_indices'.format(ListExperiences[class_order_idx][:5],list(i)[0],list(i)[1],experience_idx+1),eval_plugin.get_all_metrics()['Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000'][0])
            np.save('./checkpoints/Cumulative/{}/{}/{}/{}/Top1_Acc_Exp_1'.format(ListExperiences[class_order_idx][:5],list(i)[0],list(i)[1],experience_idx+1),eval_plugin.get_all_metrics()['Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp001'][1])
            np.save('./checkpoints/Cumulative/{}/{}/{}/{}/Top1_Acc_Exp_1_indices'.format(ListExperiences[class_order_idx][:5],list(i)[0],list(i)[1],experience_idx+1),eval_plugin.get_all_metrics()['Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp001'][0])
            np.save('./checkpoints/Cumulative/{}/{}/{}/{}/Top1_Acc_stream'.format(ListExperiences[class_order_idx][:5],list(i)[0],list(i)[1],experience_idx+1),eval_plugin.get_all_metrics()['Top1_Acc_Stream/eval_phase/test_stream/Task000'][1])
            np.save('./checkpoints/Cumulative/{}/{}/{}/{}/Top1_Acc_steam_indices'.format(ListExperiences[class_order_idx][:5],list(i)[0],list(i)[1],experience_idx+1),eval_plugin.get_all_metrics()['Top1_Acc_Stream/eval_phase/test_stream/Task000'][0])
    if class_order_idx ==0 :
        break

results.to_csv("./checkpoints/model_accuracies.csv",index=False)
```
